Use document modification date to determine dossier end date.

njohner opened the pull request:
We adapt the `earliest_possible_end_date` method to use the document modification date instead of the document date in the determination of the dossier end date when resolving a dossier. We do this because the document date will not be used to reflect modifications to the documents anymore, instead it will be set by the user.

We also introduce a new marker interface `IAutomaticallyGeneratedDocument` to identify documents that were automatically generated, as opposed to user-generated. We add this as base class for `IDossierTasksPdfMarker` and `IDossierJournalPdfMarker`, identifying the two PDFs that potentially get generated when the dossier is resolved. 

Note that `earliest_possible_end_date` is also used when determining whether a dossier has a valid end date.

resolves #4555

njohner commented:
Ok I updated this PR to make use of the newly defined changed field (see #4962).
 
`IAutomaticallyGeneratedDocument` as baseclass for `IDossierTasksPdfMarker` and `IDossierJournalPdfMarker` would probably require an upgrade step??

njohner commented on opengever/base/behaviors/changed.py:
```diff
@@ -48,8 +48,6 @@ def _set_changed(self, value):
         assert isinstance(value, DateTime), "changed should be a zope datetime"
-        # Make sure we have a timezone aware datetime
-        assert value.timezoneNaive() is False, "changed should be timezone aware"
 
```
asserting this was a bad idea. Actually zope DateTime does not distinguish between timezone naive and timezone aware, and it does not always set the `_timezone_naive` attribute. So having this is actually a pain. In tests it would require things like:
```
date = DateTime(2018, 3, 1)
date._timezone_naive = False
IChanged(document).changed = date
```

deiferni commented on opengever/document/interfaces.py:
```diff
@@ -52,6 +52,19 @@ class IDocumentType(Interface):
+    and documents added automatically by GEVER."""
+
+
+class IDossierTasksPdfMarker(IAutomaticallyGeneratedDocument):
```
pedantic naming suggestion: Pdf -> PDF

njohner replied:
Sure. I also find Pdf ugly, but the idea was to make word separation clearer...

deiferni commented on opengever/dossier/base.py:
```diff
@@ -337,7 +339,10 @@ def earliest_possible_end_date(self):
 
         for document_brain in document_brains:
-            dates.append(document_brain.document_date)
+            document = document_brain.getObject()
```
this may cause issues for larger dossiers as loading the object will be a bit more expensive. You could implement this for the brain by using the `object_provides` metadata and sacrifice the object orientation 🤔.

njohner commented:
I'll also add an upgrade step for the two interfaces that were moved (`IDossierTasksPdfMarker` and `IDossierJournalPdfMarker`)

njohner commented:
Ok I incorporated your comments and added an upgrade step to replace the old interfaces with the new ones on the objects. I also added safeguards in the upgrade, so that it does not fail if the interfaces have been removed from the code.
I think this is ready to ship.

njohner commented:
PS: I'll squash the last two commits into the others once you have reviewed this. I left them separate for now to make it easier to track the changes

deiferni commented on opengever/dossier/base.py:
```diff
@@ -337,11 +337,18 @@ def earliest_possible_end_date(self):
+                ],
+            })
+
+        autogenerated_document_uid = [document.UID for document in autogenerated_document_brains]
```
this could/should be a set, as it may contain "many" documents and you're doing "many" `in` checks against it.

njohner replied:
I did not know this was more efficient?
Is it worth making both `autogenerated_document_uid` and `document_brains` sets and then use `set.difference`?

deiferni commented on opengever/core/upgrades/20181101074702_add_automatically_generated_document_interface/upgrade.py:
```diff
@@ -0,0 +1,54 @@
+            for obj in self.objects(query, message):
+                alsoProvides(obj, new_interface)
+                noLongerProvides(obj, old_interface)
+                catalog.reindexObject(obj, idxs=['object_provides'])
```
you should no longer use `catalog.reindexObject`, but always `object.reindexObject` in upgrades. Only with `object.reindexObject` it is guaranteed that SOLR also reindexes the object. Otherwise you will only reindex `portal_catalog`.

_This unfortunately at the moment comes at the cost of no longer being able to not rebuild metadata._

njohner replied:
Yes right, I knew that...

njohner commented:
Ok this is ready to ship. I incorporated the last two comments from @deiferni . I also renamed `IAutomaticallyGeneratedDocument` to `IAutoGeneratedDocument` as per @lukasgraf 's request.

As I rewrote history to clean up the commits, here is the diff with the previous state:

```
diff --git a/opengever/core/upgrades/20181101074702_add_automatically_generated_document_interface/upgrade.py b/opengever/core/upgrades/20181101074702_add_automatically_generated_document_interface/upgrade.py
index 17866767f..7223ab56d 100644
--- a/opengever/core/upgrades/20181101074702_add_automatically_generated_document_interface/upgrade.py
+++ b/opengever/core/upgrades/20181101074702_add_automatically_generated_document_interface/upgrade.py
@@ -51,4 +51,4 @@ class AddAutomaticallyGeneratedDocumentInterface(UpgradeStep):
             for obj in self.objects(query, message):
                 alsoProvides(obj, new_interface)
                 noLongerProvides(obj, old_interface)
-                catalog.reindexObject(obj, idxs=['object_provides'])
+                obj.reindexObject(idxs=['object_provides'])
diff --git a/opengever/document/base.py b/opengever/document/base.py
index 8c19cea9d..28bcd5b1c 100644
--- a/opengever/document/base.py
+++ b/opengever/document/base.py
@@ -2,7 +2,7 @@ from Acquisition import aq_base
 from Acquisition import aq_inner
 from Acquisition import aq_parent
 from opengever.base.browser.helper import get_css_class
-from opengever.document.interfaces import IAutomaticallyGeneratedDocument
+from opengever.document.interfaces import IAutoGeneratedDocument
 from opengever.dossier.behaviors.dossier import IDossierMarker
 from opengever.dossier.dossiertemplate.behaviors import IDossierTemplateMarker
 from opengever.dossier.templatefolder.interfaces import ITemplateFolder
@@ -109,7 +109,7 @@ class BaseDocumentMixin(object):
         return IProposal.providedBy(parent)
 
     def is_automatically_generated(self):
-        return IAutomaticallyGeneratedDocument.providedBy(self)
+        return IAutoGeneratedDocument.providedBy(self)
 
     def related_items(self):
         raise NotImplementedError
diff --git a/opengever/document/interfaces.py b/opengever/document/interfaces.py
index 431293da9..b50053046 100644
--- a/opengever/document/interfaces.py
+++ b/opengever/document/interfaces.py
@@ -52,16 +52,16 @@ class IDocumentType(Interface):
     )
 
 
-class IAutomaticallyGeneratedDocument(Interface):
+class IAutoGeneratedDocument(Interface):
     """A marker interface to distinguish between user-added documents
     and documents added automatically by GEVER."""
 
 
-class IDossierTasksPDFMarker(IAutomaticallyGeneratedDocument):
+class IDossierTasksPDFMarker(IAutoGeneratedDocument):
     """Marker Interface for dossier tasks list document."""
 
 
-class IDossierJournalPDFMarker(IAutomaticallyGeneratedDocument):
+class IDossierJournalPDFMarker(IAutoGeneratedDocument):
     """Marker Interface for dossier journal document."""
 
 
diff --git a/opengever/dossier/base.py b/opengever/dossier/base.py
index dbf8825a3..b099d3801 100644
--- a/opengever/dossier/base.py
+++ b/opengever/dossier/base.py
@@ -340,11 +340,11 @@ class DossierContainer(Container):
         autogenerated_document_brains = catalog({
             'path': '/'.join(self.getPhysicalPath()),
             'object_provides': [
-                'opengever.document.interfaces.IAutomaticallyGeneratedDocument',
+                'opengever.document.interfaces.IAutoGeneratedDocument',
                 ],
             })
 
-        autogenerated_document_uid = [document.UID for document in autogenerated_document_brains]
+        autogenerated_document_uid = set(document.UID for document in autogenerated_document_brains)
         filtered_brains = filter(lambda document: document.UID not in autogenerated_document_uid,
                                  document_brains)
```

njohner commented:
wait for https://github.com/4teamwork/opengever.core/pull/5048 to be merged, then check the tests again here

njohner commented:
Last minute changes. The main thing is not to cast zope `DateTime` to python `datetime` in `_convert_to_date`. We had added that as `changed` was a zope `DateTime`, but this is not longer the case. Moreover this is dangerous, as we were using `asdatetime` here. Instead we make sure in the `changed` getter to return a python `datetime`

The rest is changes in the tests

```
diff --git a/opengever/base/behaviors/changed.py b/opengever/base/behaviors/changed.py
index baac58eb2..ca2004196 100644
--- a/opengever/base/behaviors/changed.py
+++ b/opengever/base/behaviors/changed.py
@@ -42,6 +42,8 @@ class Changed(object):
         # https://github.com/4teamwork/opengever.core/issues/4988
         if self.context.changed is None:
             return as_utc(self.context.modified().asdatetime())
+        if isinstance(self.context.changed, DateTime):
+            return as_utc(self.context.changed.asdatetime())
         return self.context.changed
 
     def _set_changed(self, value):
diff --git a/opengever/dossier/base.py b/opengever/dossier/base.py
index 7f9366bf4..232c19198 100644
--- a/opengever/dossier/base.py
+++ b/opengever/dossier/base.py
@@ -460,9 +460,6 @@ class DossierContainer(Container):
         if isinstance(datetime_obj, datetime):
             return datetime_obj.date()
 
-        if isinstance(datetime_obj, DateTime):
-            return datetime_obj.asdatetime().date()
-
         # It is already a date-object
         return datetime_obj
 
diff --git a/opengever/dossier/tests/test_archiver.py b/opengever/dossier/tests/test_archiver.py
index 99c0e830c..4832e36a0 100644
--- a/opengever/dossier/tests/test_archiver.py
+++ b/opengever/dossier/tests/test_archiver.py
@@ -1,6 +1,5 @@
 from datetime import date
 from datetime import datetime
-from DateTime import DateTime
 from ftw.builder import Builder
 from ftw.builder import create
 from ftw.testbrowser import browsing
@@ -31,6 +30,7 @@ from opengever.dossier.interfaces import IDossierArchiver
 from opengever.testing import IntegrationTestCase
 from zope.interface import Invalid
 from zope.interface.verify import verifyClass
+import pytz
 
 
 class TestArchiver(IntegrationTestCase):
@@ -249,7 +249,8 @@ class TestArchiveFormDefaults(IntegrationTestCase):
         self.login(self.dossier_responsible, browser)
         IDossier(self.dossier).end = date(2021, 1, 1)
         self.dossier.reindexObject(idxs=['end'])
-        IChanged(self.subdocument).changed = DateTime(2021, 1, 2)
+
+        IChanged(self.subdocument).changed = datetime(2021, 1, 2, tzinfo=pytz.utc)
         self.subdocument.reindexObject(idxs=['changed'])
         browser.open(self.dossier, view='transition-archive')
         self.assertEqual(date(2021, 1, 2),
@@ -263,7 +264,7 @@ class TestArchiveFormDefaults(IntegrationTestCase):
         self.login(self.dossier_responsible, browser)
         IDossier(self.dossier).end = date(2021, 2, 2)
         self.dossier.reindexObject(idxs=['end'])
-        IChanged(self.subdocument).changed = DateTime(2021, 1, 2)
+        IChanged(self.subdocument).changed = datetime(2021, 1, 2, tzinfo=pytz.utc)
         self.subdocument.reindexObject(idxs=['changed'])
         browser.open(self.dossier, view='transition-archive')
         self.assertEqual(date(2021, 2, 2),
diff --git a/opengever/dossier/tests/test_base.py b/opengever/dossier/tests/test_base.py
index cb26362cb..2dea985bb 100644
--- a/opengever/dossier/tests/test_base.py
+++ b/opengever/dossier/tests/test_base.py
@@ -1,6 +1,5 @@
 from datetime import date
 from datetime import datetime
-from DateTime import DateTime
 from ftw.builder import Builder
 from ftw.builder import create
 from ftw.testbrowser import browsing
@@ -15,6 +14,7 @@ from plone import api
 from plone.registry.interfaces import IRegistry
 from zope.component import getUtility
 from zope.interface import alsoProvides
+import pytz
 
 
 class TestDossierContainer(IntegrationTestCase):
@@ -224,7 +224,7 @@ class TestDateCalculations(IntegrationTestCase):
 
     def test_earliest_possible_is_last_modified_document_modification_date(self):
         self.login(self.dossier_responsible)
-        IChanged(self.document).changed = DateTime(2021, 1, 22)
+        IChanged(self.document).changed = datetime(2021, 1, 22, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
         self.assertEquals(date(2021, 1, 22),
                           self.dossier.earliest_possible_end_date())
@@ -233,7 +233,7 @@ class TestDateCalculations(IntegrationTestCase):
         self.login(self.dossier_responsible)
         IDossier(self.dossier).end = date(2021, 1, 21)
         self.dossier.reindexObject(idxs=['end'])
-        IChanged(self.document).changed = DateTime(2021, 1, 22)
+        IChanged(self.document).changed = datetime(2021, 1, 22, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
 
         self.assertEquals(date(2021, 1, 22),
@@ -248,13 +248,13 @@ class TestDateCalculations(IntegrationTestCase):
     def test_earliest_possible_is_latest_of_dossiers_end_dates_and_document_modificiation_dates(self):
         self.login(self.dossier_responsible)
 
-        IChanged(self.document).changed = DateTime(2020, 1, 1)
+        IChanged(self.document).changed = datetime(2020, 1, 1, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
         IDossier(self.subdossier).end = date(2020, 2, 2)
         self.subdossier.reindexObject(idxs=['end'])
         self.assertEquals(date(2020, 2, 2), self.dossier.earliest_possible_end_date())
 
-        IChanged(self.document).changed = DateTime(2020, 3, 3)
+        IChanged(self.document).changed = datetime(2020, 3, 3, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
         self.assertEquals(date(2020, 3, 3), self.dossier.earliest_possible_end_date())
 
@@ -292,21 +292,21 @@ class TestDateCalculations(IntegrationTestCase):
     def test_dossier_end_can_be_later_than_document_date(self):
         self.login(self.dossier_responsible)
         IDossier(self.dossier).end = date(2020, 2, 2)
-        IChanged(self.document).changed = DateTime(2020, 1, 1)
+        IChanged(self.document).changed = datetime(2020, 1, 1, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
         self.assertTrue(self.dossier.has_valid_enddate())
 
     def test_dossier_end_can_be_equal_to_document_modification_date(self):
         self.login(self.dossier_responsible)
         IDossier(self.dossier).end = date(2020, 1, 1)
-        IChanged(self.document).changed = DateTime(2020, 1, 1)
+        IChanged(self.document).changed = datetime(2020, 1, 1, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
         self.assertTrue(self.dossier.has_valid_enddate())
 
     def test_dossier_end_cannot_be_earlier_to_document_date(self):
         self.login(self.dossier_responsible)
         IDossier(self.dossier).end = date(2020, 1, 1)
-        IChanged(self.document).changed = DateTime(2020, 2, 2)
+        IChanged(self.document).changed = datetime(2020, 2, 2, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
         self.assertFalse(self.dossier.has_valid_enddate())
 
@@ -337,7 +337,7 @@ class TestDateCalculations(IntegrationTestCase):
     def test_earliest_possible_can_handle_datetime_objs(self):
         self.login(self.dossier_responsible)
 
-        IChanged(self.document).changed = DateTime(2020, 1, 1, 10, 10)
+        IChanged(self.document).changed = datetime(2020, 1, 1, 10, 10, tzinfo=pytz.utc)
         self.document.reindexObject(idxs=['changed'])
         IDossier(self.subdossier).end = datetime(2020, 2, 2, 10, 10)
         self.subdossier.reindexObject(idxs=['end'])
```

njohner commented:
Ok that was dumb 😓 . When calculating the `earliest_possible_end_date` for a dossier, we get the `changed` dates from the brains, hence we do not go through the `IChanged` behavior getter. As `changed` can be a zope `DateTime` on installations that were updated before we switched the `changed` field to a python datetime, this will lead to dates that are not comparable if we do not run the upgrade that fixes the `changed` dates (which is deferrable). So we make sure to cast those dates to python `date` here
